--- a/src/freedreno/vulkan/tu_knl_kgsl.cc
+++ b/src/freedreno/vulkan/tu_knl_kgsl.cc
@@ -890,121 +890,87 @@
    }
 }
 
-#define kgsl_syncobj_foreach_state(syncobjs, filter) \
-   for (uint32_t i = 0; sync = syncobjs[i], i < count; i++) \
-      if (sync->state == filter)
-
-static VkResult
-kgsl_syncobj_wait_any(struct tu_device* device, struct kgsl_syncobj **syncobjs, uint32_t count, uint64_t abs_timeout_ns)
+static VkResult
+kgsl_syncobj_wait_any(struct tu_device *device, struct kgsl_syncobj **syncobjs,
+                      uint32_t count, uint64_t abs_timeout_ns)
 {
    if (count == 0)
       return VK_TIMEOUT;
-   else if (count == 1)
+   if (count == 1)
       return kgsl_syncobj_wait(device, syncobjs[0], abs_timeout_ns);
 
-   uint32_t num_fds = 0;
    struct tu_queue *queue = NULL;
-   struct kgsl_syncobj *sync = NULL;
-
-   /* Simple case, we already have a signal one */
-   kgsl_syncobj_foreach_state(syncobjs, KGSL_SYNCOBJ_STATE_SIGNALED)
-      return VK_SUCCESS;
-
-   kgsl_syncobj_foreach_state(syncobjs, KGSL_SYNCOBJ_STATE_FD)
-      num_fds++;
-
-   /* If we have TS from different queues we cannot compare them and would
-    * have to convert them into FDs
-    */
-   bool convert_ts_to_fd = false;
-   kgsl_syncobj_foreach_state(syncobjs, KGSL_SYNCOBJ_STATE_TS) {
-      if (queue != NULL && sync->queue != queue) {
-         convert_ts_to_fd = true;
-         break;
-      }
-      queue = sync->queue;
-   }
-
-   /* If we have no FD nor TS syncobjs then we can return immediately */
-   if (num_fds == 0 && queue == NULL)
-      return VK_TIMEOUT;
-
-   VkResult result = VK_TIMEOUT;
-
-   struct u_vector poll_fds = { 0 };
-   uint32_t lowest_timestamp = 0;
-
-   if (convert_ts_to_fd || num_fds > 0)
-      u_vector_init(&poll_fds, 4, sizeof(struct pollfd));
-
-   if (convert_ts_to_fd) {
-      kgsl_syncobj_foreach_state(syncobjs, KGSL_SYNCOBJ_STATE_TS) {
-         struct pollfd *poll_fd = (struct pollfd *) u_vector_add(&poll_fds);
-         poll_fd->fd = timestamp_to_fd(sync->queue, sync->timestamp);
-         poll_fd->events = POLLIN;
-      }
-   } else {
-      /* TSs could be merged by finding the one with the lowest timestamp */
-      bool first_ts = true;
-      kgsl_syncobj_foreach_state(syncobjs, KGSL_SYNCOBJ_STATE_TS) {
-         if (first_ts || timestamp_cmp(sync->timestamp, lowest_timestamp)) {
-            first_ts = false;
+   uint32_t num_fds = 0, num_ts = 0, lowest_timestamp = 0;
+   bool multiple_queues = false;
+   for (uint32_t i = 0; i < count; ++i) {
+      const struct kgsl_syncobj *sync = syncobjs[i];
+      if (sync->state == KGSL_SYNCOBJ_STATE_SIGNALED)
+         return VK_SUCCESS;
+      if (sync->state == KGSL_SYNCOBJ_STATE_FD) {
+         num_fds++;
+      } else if (sync->state == KGSL_SYNCOBJ_STATE_TS) {
+         if (queue == NULL) {
+            queue = sync->queue;
             lowest_timestamp = sync->timestamp;
+         } else if (queue != sync->queue) {
+            multiple_queues = true;
+         } else {
+            lowest_timestamp = min_ts(lowest_timestamp, sync->timestamp);
          }
-      }
-
-      if (num_fds) {
-         struct pollfd *poll_fd = (struct pollfd *) u_vector_add(&poll_fds);
-         poll_fd->fd = timestamp_to_fd(queue, lowest_timestamp);
-         poll_fd->events = POLLIN;
-      }
-   }
-
-   if (num_fds) {
-      kgsl_syncobj_foreach_state(syncobjs, KGSL_SYNCOBJ_STATE_FD) {
-         struct pollfd *poll_fd = (struct pollfd *) u_vector_add(&poll_fds);
-         poll_fd->fd = sync->fd;
-         poll_fd->events = POLLIN;
-      }
-   }
-
-   if (u_vector_length(&poll_fds) == 0) {
-      result = wait_timestamp_safe(device->fd, queue->msm_queue_id,
-                                   lowest_timestamp, MIN2(abs_timeout_ns, INT64_MAX));
-   } else {
-      int ret, i;
-
-      struct pollfd *fds = (struct pollfd *) poll_fds.data;
-      uint32_t fds_count = u_vector_length(&poll_fds);
-      do {
-         ret = poll(fds, fds_count, get_relative_ms(abs_timeout_ns));
-         if (ret > 0) {
-            for (i = 0; i < fds_count; i++) {
-               if (fds[i].revents & (POLLERR | POLLNVAL)) {
-                  errno = EINVAL;
-                  ret = -1;
-                  break;
-               }
+         num_ts++;
+      }
+   }
+   if (num_fds == 0) {
+      if (num_ts == 0)
+         return VK_TIMEOUT;
+      if (!multiple_queues)
+         return wait_timestamp_safe(device->fd, queue->msm_queue_id,
+                                    lowest_timestamp, abs_timeout_ns);
+   }
+
+   struct pollfd *fds = (struct pollfd *) calloc(count, sizeof(*fds));
+   if (fds == NULL)
+      return VK_ERROR_OUT_OF_HOST_MEMORY;
+   uint32_t owned_fds = 0, fds_count = 0;
+   VkResult result = VK_ERROR_UNKNOWN;
+   for (uint32_t i = 0; i < count; ++i) {
+      const struct kgsl_syncobj *sync = syncobjs[i];
+      if (sync->state != KGSL_SYNCOBJ_STATE_TS)
+         continue;
+      int fd = timestamp_to_fd(sync->queue, sync->timestamp);
+      if (fd < 0)
+         goto finish;
+      fds[fds_count++] = { fd, POLLIN, 0 };
+      owned_fds++;
+   }
+   for (uint32_t i = 0; i < count; ++i) {
+      if (syncobjs[i]->state == KGSL_SYNCOBJ_STATE_FD)
+         fds[fds_count++] = { syncobjs[i]->fd, POLLIN, 0 };
+   }
+
+   while (true) {
+      int ret = poll(fds, fds_count, get_relative_ms(abs_timeout_ns));
+      if (ret < 0 && (errno == EINTR || errno == EAGAIN))
+         continue;
+      if (ret == 0) {
+         result = VK_TIMEOUT;
+      } else if (ret > 0) {
+         for (uint32_t i = 0; i < fds_count; ++i) {
+            if (fds[i].revents & (POLLERR | POLLNVAL | POLLHUP)) {
+               result = VK_ERROR_UNKNOWN;
+               goto finish;
             }
-            break;
-         } else if (ret == 0) {
-            errno = ETIME;
-            break;
+            if (fds[i].revents & POLLIN)
+               result = VK_SUCCESS;
          }
-      } while (ret == -1 && (errno == EINTR || errno == EAGAIN));
-
-      for (uint32_t i = 0; i < fds_count - num_fds; i++)
-         close(fds[i].fd);
-
-      if (ret != 0) {
-         assert(errno == ETIME);
-         result = VK_TIMEOUT;
-      } else {
-         result = VK_SUCCESS;
-      }
-   }
-
-   u_vector_finish(&poll_fds);
+      }
+      break;
+   }
+
+finish:
+   for (uint32_t i = 0; i < owned_fds; ++i)
+      close(fds[i].fd);
+   free(fds);
    return result;
 }
 
